src/controllers/savedBooksController.js

```diff
@@ -0,0 +1,172 @@
+const { StatusCodes } = require('http-status-codes');
+const { BadRequestError, NotFoundError } = require('../errors');
+const SavedBooks = require('../models/SavedBooks');
+const Book = require('../models/Book');
+
+const addBookToSavedBooks = async (req, res, next) => {
+  try {
+    const { userId } = req.user;
+    const { isbn10, isbn13 } = req.body;
+
+    if (!isbn10 && !isbn13) {
+      return next(
+        new BadRequestError(
+          'At least one ISBN (ISBN-10 or ISBN-13) must be provided.'
+        )
+      );
+    }
+
+    let savedBooks = await SavedBooks.findOne({ user: userId });
+
+    if (!savedBooks) {
+      savedBooks = new SavedBooks({
+        user: userId,
+        books: [],
+      });
+    }
+
+    const bookExists = savedBooks.books.some(
+      (book) => book.isbn10 === isbn10 || book.isbn13 === isbn13
+    );
+
+    if (bookExists) {
+      return res
+        .status(StatusCodes.BAD_REQUEST)
+        .json({ msg: 'This book is already saved.' });
+    }
+
+    const newBook = {
+      isbn10,
+      isbn13,
+      addedAt: new Date(),
+    };
+
+    savedBooks.books.push(newBook);
+
+    await savedBooks.save();
+
+    res
+      .status(StatusCodes.CREATED)
+      .json({ msg: 'Book added successfully', savedBooks });
+  } catch (error) {
+    console.error('Error adding book:', error);
+    next(error);
+  }
+};
+
+const getSavedBooks = async (req, res, next) => {
+  try {
+    const { userId } = req.user;
+    const { sort = '-addedAt' } = req.query;
+    const savedBooks = await SavedBooks.findOne({ user: userId });
+
+    if (!savedBooks || savedBooks.books.length === 0) {
+      return res.status(StatusCodes.OK).json({
+        msg: 'No saved books found for this user.',
+        savedBooks: {
+          books: [],
+        },
+      });
+    }
+
+    const booksWithDetails = [];
+
```

Comment:
using for loop with await can cause sequential execution, you can parallelize it with Primse.all:
const booksWithDetails = await Promise.all(
  savedBooks.books.map(async (savedBook) => {
    const bookDetails = await Book.find({
      $or: [{ isbn13: savedBook.isbn13 }, { isbn10: savedBook.isbn10 }],
    });
    
    Not a needed fix, but this is good to know.